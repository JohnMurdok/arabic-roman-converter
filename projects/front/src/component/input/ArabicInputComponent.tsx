import React, { useContext } from 'react';
import { useSetState } from 'react-use';
import { useDebounce } from 'react-use';
import { TextField } from '@mui/material';
import DEFAULT_INPUT_PROPS, { DEBOUNCE_TIME } from '../../constant/input';
import numberConverter from '../../service/numberConverter';
import context from '../../context';
import sseService from '../../service/sse';
import EventEnum from '../../enum/event';

const ARABIC_INPUT_PROPS = {
    ...DEFAULT_INPUT_PROPS,
    type: 'number',
    label: 'Arabic number',
    inputProps: {
        min: 1,
        step: 1,
    }
};

const NUMBER_REG_EXP = /^\d+$/;
interface IMessageEvent {
    romanNumber: string;
}

const ArabicInputComponent = () => {
    const {
        arabicNumber,
        setRomanNumber,
        setArabicNumber,
    } = useContext(context);
    const [state, setState] = useSetState({
        error: false,
        arabicNumber,
    });

    const onChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
        ev.preventDefault();
        const { value } = ev.target;
        const newValue = Number(value);
        if (!value.match(NUMBER_REG_EXP) || newValue <= 0) {
            setState({
                arabicNumber: newValue,
                error: true,
            });
            return;
        }
        setState({
            arabicNumber: newValue,
            error: false,
        });
        setArabicNumber(newValue);
    };

    useDebounce(async () => {
        if (!arabicNumber || state.arabicNumber !== arabicNumber || state.error) {
            return;
        }

        sseService.registerEvent(EventEnum.CONVERT_TO_ROMAN, (ev: IMessageEvent) => {
            setRomanNumber(ev.romanNumber);
        });

        await numberConverter.convertToRoman(arabicNumber, sseService.clientUuid);
    }, DEBOUNCE_TIME,[arabicNumber, state]);


    return (
        <TextField
            data-testid={'arabic-input'}
            error={!!state.error}
            {...ARABIC_INPUT_PROPS} onChange={onChange} value={arabicNumber} />
    );
}

export default ArabicInputComponent;
