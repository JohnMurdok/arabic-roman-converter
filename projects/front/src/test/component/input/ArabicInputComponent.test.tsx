import { render, screen } from '@testing-library/react';
import ArabicInputComponent from '../../../component/input/ArabicInputComponent';
jest.mock('../../../service/sse', () => ({
    clientUuid: 'uuid',
    registerEvent: jest.fn((event, cb) => cb({
        arabicNumber: 1,
    })),
}));

describe('[COMPONENT] ArabicInputComponent', () => {
    test('matches snapshot', async () => {
        render(<ArabicInputComponent />);
        const textField = await screen.findByTestId('arabic-input');
        expect(textField).toMatchSnapshot();
    });
});
