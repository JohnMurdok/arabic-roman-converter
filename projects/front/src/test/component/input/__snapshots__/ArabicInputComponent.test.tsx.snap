// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[COMPONENT] ArabicInputComponent matches snapshot 1`] = `
<div
  class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
  data-testid="arabic-input"
>
  <label
    class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink="true"
    for=":r0:"
    id=":r0:-label"
  >
    Arabic number
    <span
      aria-hidden="true"
      class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
    >
       
      *
    </span>
  </label>
  <div
    class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
  >
    <input
      aria-invalid="false"
      class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
      id=":r0:"
      min="1"
      required=""
      step="1"
      type="number"
      value="1"
    />
    <fieldset
      aria-hidden="true"
      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="css-1in441m"
      >
        <span>
          Arabic number
           
          *
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
